staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/escaping.go

```diff
@@ -36,6 +36,57 @@ var celReservedSymbols = sets.NewString(
 // expandMatcher matches the escape sequence, characters that are escaped, and characters that are unsupported
 var expandMatcher = regexp.MustCompile(`(__|[-./]|[^a-zA-Z0-9-./_])`)
 
+// newCharacterFilter returns a boolean array to indicate the allowed characters
+func newCharacterFilter(characters string) []bool {
+	maxChar := 0
+	for _, c := range characters {
+		if maxChar < int(c) {
+			maxChar = int(c)
+		}
+	}
+	filter := make([]bool, maxChar+1)
+
+	for _, c := range characters {
+		filter[int(c)] = true
+	}
+
+	return filter
+}
+
+type escapeCheck struct {
+	canSkipRegex     bool
+	invalidCharFound bool
+}
+
+// skipRegexCheck checks if escape would be skipped.
+// if invalidCharFound is true, it must have invalid character; if invalidCharFound is false, not sure if it has invalid character or not
+func skipRegexCheck(ident string) escapeCheck {
+	escapeCheck := escapeCheck{canSkipRegex: true, invalidCharFound: false}
+	// skip escape if possible
+	previous_underscore := false
+	for _, c := range ident {
+		if c == '/' || c == '-' || c == '.' {
+			escapeCheck.canSkipRegex = false
+			return escapeCheck
+		}
+		intc := int(c)
+		if intc < 0 || intc >= len(validCharacterFilter) || !validCharacterFilter[intc] {
```

Comment:
It seems simpler to omit the `intc < 0 || intc >= len(validCharacterFilter)` check and just keep a set of the allowed chars (which simplifies `newCharacterFilter`).

I would also expect this to be marginally faster for valid identifiers (which will be in the set so will pass all these checks), and optimizing for valid identifiers seems right here.

Reply:
An alternative to the set would be to do everything via comparisons, e.g.:

I'm fine with this. An alternative would be to do all the checks using ranges, e.g. `(48 <= intc && intc <= 57) || (65 <= intc && intc <= 90) || (97 <= intc && intc <= 122)`.

Since you have a benchmark, doing whichever is faster seems reasonable.

Reply:
Updated the performance results for those alternative solutions in PR description. Looks like the current way gets more performance boost